File: 0086-partition-list/0086-partition-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* lessDummy = new ListNode(-1); 
        ListNode* greaterDummy = new ListNode(-1); 

        ListNode* less = lessDummy;
        ListNode* greater = greaterDummy; 

        ListNode* curr = head; 
        while(curr){
            if(curr->val < x){
                less->next = curr; 
                less = less->next; 
            } else{
                greater->next = curr; 
                greater = greater->next; 
            }
            curr = curr->next; 
        }

        greater->next = nullptr; 
        less->next = greaterDummy->next; 

        return lessDummy->next; 
    }
};

File: 0086-partition-list/0086-partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode smallDummy = new ListNode(-1); 
        ListNode bigDummy = new ListNode(-1); 
        
        ListNode smallCurr = smallDummy; 
        ListNode bigCurr = bigDummy; 

        while(head != null){
            int val = head.val; 
            if(val < x){
                smallCurr.next = new ListNode(val); 
                smallCurr = smallCurr.next; 
            } else{
                bigCurr.next = new ListNode(val); 
                bigCurr = bigCurr.next; 
            }
            head = head.next; 
        }

        smallCurr.next = bigDummy.next; 

        
        return smallDummy.next; 
    }
}